README
Alchemy/WebMail version 1.0
===========================

 Alchemy WebMail is a Web based mod_perl e-mail client. It relies on
 a database for storing user application data and an IMAP server for
 managing the user's e-mail. 

 Alchemy Webmail should work with any database for which a DBI layer
 exists, the preferred database is PostgreSQL. 

INSTALLATION

 Ensure that you have Apache 2.x and mod_perl 2.x.

 To install this module type the following:

   perl Build.PL
   ./Build
   ./Build test
   ./Build install

 Database Configuration:

   Load the appropriate schema file into your database according to your
   particular database's methods.  Note that the schema files contain
   only the table and sequence shema, they do not contain any
   authorization information for the database. It is up to the installer
   to set these to their preference.

 Apache Configuration:

   Configure your apache installation for mod_perl and Alchemy WebMail,
   see the Webmail.conf for an example configuration.  
   
   "perldoc Alchemy::Webmail" details all of the configuration options.

   Copy or create templates for your installation of WebMail. The
   default templates are located in the templates directory of the
   source distribution.

   After these steps have been completed you have a dandy web based
   e-mail application running that your users should like, if they know
   what's good for them. ;)

DEPENDENCIES

 This module requires these other modules and libraries:

	mod_perl 2.x 

	KrKit
	Crypt::CBC
		Crypt::Blowfish or another cipher that works with Crypt::CBC
	Mail::IMAPClient
	MIME::Entity
	MIME::Parser

COPYRIGHT AND LICENSE

 Copyright (c) 2003-2010 Nicholas Studt. All rights reserved.

 You may distribute under the terms of either the GNU General Public
 License or the Artistic License, as specified in the Perl README file.



Changes
Revision history for Perl extension Alchemy::WebMail.

-[ Todo ]--------------------------------------------------------------

	- Fix the address book to work in both a popup and in the full
	  screen. Both should have the same features, but one should just
	  have a different template.
		- Groups, move rename into edit.

	- Add an icon for attachments in the folder listing, how do I tell ?

	- Make the default templates prettier.
	- Document install some more.
	- Exploit the Cyrus/SASL to do real folders.
	- Build the default SQL creation scripts for Mysql and Oracle.

Requested Features:
	- Search mail.
	- Display attachment icon in main mail listing.
	- Allow multiple attachments with the same name (seen from forwards)
	- Fix the stale attachments from compose...on close script?

-[ Revision History ]--------------------------------------------------

1.0.1 Mon May 10 21:25:28 CDT 2010
	- Fixed the bug that allowed a blank subject to be displayed
	  rendering it impossible to open the message from the list view.

1.0 Sat Jan 16 16:15:41 CST 2010
	- Converted from Mail::Cclient to Mail::IMAPClient.
	- Moved to Module::Build over ExtUtils::MakeMaker, goodbye make.
	- Fixed a bug that made HTML only mail display as plain text.
	- Fixed the Folder creation to no allow leading/trailing spaces.
	- Removed dta css class from use.

0.77 Mon Dec 15 21:32:54 CST 2008
	- Last working in AddressBook.

0.76 Wed Jun  4 20:48:54 CDT 2008
	- Modified encoding to not quote the default message if it is
	  html... This probably needs some more thought.
	- Moved from using multiple to single inheritance.	
	- Roles now save and recover from draft correctly.

0.75 Thu Apr  5 13:55:15 CDT 2007
	- Sunflower patch, fixed dropping plain text attachments in
	  multipart messages.
	- Sunflower patch, sort order on attachments. ( Now list in order it
	  is in the email )
	- Sunflower patch, ISO handling in subject, from line, email.

0.74 Fri Sep  2 09:00:50 CDT 2005
	- Moved Webmail.conf to httpd.conf.
	- Updating to the Apache2::* naming convention for mod_perl 2.000001

0.73 Thu Dec 30 20:49:14 CST 2004
	- Fixed the body_fh problem. Requires KrKit 0.14

0.72 Thu Dec 30 10:24:56 CST 2004
	- Fixed reply to in the sentmail folder.
	- Adjusted addsig to add the signature where the cursor is at in the
	  message, or if not in the message textarea then add to the end.
	  The method comes from http://www.alexking.org/blog/2003/06/02/
	- Added the ability to add a address from the popup addressbook page.
	- Dates in the listing now are formatable via DateTime_Format.
	
0.71 Mon Dec 27 11:48:10 CST 2004
	- Fixing the authentication handler to not popup.

0.70 Sun Dec 26 18:04:25 CST 2004
	- porting to mod_perl 2

0.63 Thu Oct  7 08:13:34 CDT 2004
	- Change counting to require a flag if we want to keep it.

0.63 Fri Aug 20 20:54:58 CDT 2004
	- Fixing the folder sort order problem?

0.62 Sat Jul 31 21:26:12 CDT 2004
	- Moved to using ht_checkbox.
	- Added the where clause to sql_update

0.61 Thu Jul 22 21:39:25 CDT 2004
	- Spelling error fix in the address book addition errors.

0.60 Thu Jun 24 22:41:58 CDT 2004
	- Made the email in the address book compose.

0.59 Thu Jun 24 08:45:43 CDT 2004
	- Changed the address book to not allow duplicate e-mails in one
	  persons address book.
	- Changed the auto add to always show the user the form to confirm
	  the information.

0.58 Wed Jun 23 21:57:45 CDT 2004
	- Finished the Groups admin.
	- Added the groups the the address book popup.
	- Added ability to grab the address from an email and put it in the
	  address book.

0.57 Tue Jun 22 21:54:53 CDT 2004
	- Fixed the title in the main message view.
	- Groups work. All implemented in the group admin save membership
	  addition and removal. All old group code has been removed with
	  prejudice, it was *bad*.

0.56 Mon Jun 21 21:34:51 CDT 2004
	- Added the ability to mask the INBOX as something else. This is an
	  option and need not be used if the system setting of INBOX is
	  useable. The option changes the displayed value only, not the
	  value used for links or folder operations.

0.55 Sun Jun 20 21:24:56 CDT 2004
	- Fixed a uninitialized value warning in Preferences::Roles.
	- Fixed login case sensitivity, now all usernames are lower cased.
	- Fixed the reply quoting.
	- Put in a temporary Inbox fix.
	- Changed the unknown.html -> 'View HTML Version of Message'
	- Do not display the .plain type. 
	- Fixed the folder sort to percolate Inbox to the top.
	- Fix the draft folder issue in do_response no need to send the
	  folder twice.

0.54b Wed Mar 31 21:44:37 CST 2004
	- Debugging release for the recent IMAP folder creation problem.

0.53 Fri Feb 13 21:44:33 CST 2004
	- Added a ht_qt around the subject.
	- Updated to the new CPAN module layout.

0.52 Fri Dec 26 16:30:59 CST 2003
	- Display of folder counts in the folder list.
	- Working on the groups form, to create a new group or edit an
	  existing group.
	- Completed group addition. Needs edit, delete, and view.

0.51 Mon Nov 24 21:44:43 CST 2003
	- Added another test for a blank subject in the folder view. 
	- Added defaults to the message view.
	- Added defaults to the responce function. 
	- Adjusted the main folder view to show the To address for the
	  sentmail folder. This only works for the "main" sentmail folder.

0.50 Sun Nov 16 15:59:08 CST 2003
	- Adjusted the mail parser to deal with nested multipart messages.
	- Adjusted the link highlighter to catch a couple more types of links.
	- Added the ability to copy a message to another folder.

0.49 Tue Oct 21 21:51:52 CDT 2003
	- Working on the groups stuff.
	- Fixed a bug in the From address display.
	- Partially fixed the "To" field in drafts.
	- Fixed a bug in the reply to code.

0.48 Sat Oct 11 18:56:29 CDT 2003
	- Fixed the SMTP sending of messages in order to ensure that the
	  message does, in fact, get sent.
	- Fixed the login redirect, forgot to send people to the proper

lib/Alchemy/WebMail/Preferences.pm
package Alchemy::WebMail::Preferences;

use strict;

use KrKit::DB;
use KrKit::HTML qw(:all);
use KrKit::SQL;
use KrKit::Validate;

use Alchemy::WebMail;

our @ISA = ('Alchemy::WebMail');

############################################################
# Functions                                                #
############################################################

#-------------------------------------------------
# $k->_checkvals($in)
#-------------------------------------------------
sub _checkvals {
	my ($k, $in) = @_;

	my @errors;

	if (! is_number($in->{reply})) {
		push(@errors, 'Select reply inclusion.');
	}

	if (! is_number($in->{delete})) {
		push(@errors, 'Select delete action.');
	}

	if (! is_text($in->{length})) {
		push(@errors, 'Select session length.');
	}

	if (! is_number($in->{count})) {
		push(@errors, 'Select folder message count.');
	}

	if (! is_text($in->{field})) {
		push(@errors, 'Select sort field.');
	}

	if (! is_number($in->{order})) {
		push(@errors, 'Select sort order.');
	}

	if (@errors) {
		return(ht_div({ 'class' => 'error' },
					ht_h(1, 'Errors:'),
					ht_ul(undef, map {ht_li(undef, $_)} @errors )));
	}

	return();
} # END $k->_checkvals

#-------------------------------------------------
# $k->_form($in)
#-------------------------------------------------
sub _form {
	my ($k, $in) = @_;

	my @opts;

	for my $sopt (split(',', $$k{'p_sessopt'})) {
		push(@opts, $sopt, ucfirst($sopt));
	}

	return(ht_form_js($$k{uri}),	
			ht_div({ 'class' => 'box' }),
			ht_table(),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Reply Included'),
				ht_td(undef,
					ht_select('reply', 1, $in, '', '', '1', 'Yes', '0', 'No'))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Delete Action'),
				ht_td(undef, ht_select('delete', 1, $in, '', '', 
										'1', 'Delete', '0', 'Move to Trash'))),
			
			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Session Length'),
				ht_td(undef, ht_select('length', 1, $in, '', '', @opts))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Message Count'),
				ht_td(undef, ht_select('count', '1', $in, '', '',
										'10', '10', '25', '25', '50', '50',
										'1000', 'All'))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Sort Field'),
				ht_td(undef, ht_select('field', '1', $in, '', '', 
										'date', 'Date', 'from', 'From',
										'subject', 'Subject', 'size', 'Size'))),
			
			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Sort Order'),
				ht_td(undef, ht_select('order', '1', $in, '', '', 
										'0', 'Ascending', '1', 'Descending'))),

			ht_tr(undef,
				ht_td({ 'colspan' => '2', 'class' => 'rshd' }, 
					ht_submit('submit', 'Save'),
					ht_submit('cancel', 'Cancel'))),

			ht_utable(),
			ht_udiv(),
			ht_uform());
} # END $k->_form

#-------------------------------------------------
# $k->do_edit($r)
#-------------------------------------------------
sub do_edit {
	my ($k, $r) = @_;

	my $in = $k->param(Apache2::Request->new($r));
	$$k{page_title}	.= 'Update Preferences';
	
	return($k->_relocate($r, $$k{rootp})) if ($in->{cancel});

	if (my @err = $k->_checkvals($in)) {

		# Look up the users id and the old values. 
		my $sth = db_query($$k{dbh}, 'get prefs', 
							'SELECT reply_include, true_delete, ',
							'session_length, fldr_showcount, fldr_sortorder,',
							'fldr_sortfield FROM wm_users WHERE id = ', 
							sql_num($$k{user_id}));
	
		my ($reply, $delete, $length, $count, $order, $field) = db_next($sth);

		db_finish($sth);

		# Set the old values if they are not already defined. 
		# Setting here rather than above lets users blank entries.
		$in->{reply} 	= $reply 	if (! defined $in->{reply});
		$in->{delete} 	= $delete 	if (! defined $in->{delete});
		$in->{count} 	= $count 	if (! defined $in->{count});
		$in->{order} 	= $order 	if (! defined $in->{order});
		$in->{field} 	= $field 	if (! defined $in->{field});
		
		$in->{length} = $k->s2hm($length) if (! defined $in->{length});

		return(($r->method eq 'POST' ? @err : ''), $k->_form($in));
	}

	my $session = $k->hm2s($in->{length});

	db_run($$k{dbh}, 'Update user prefrences', 
			sql_update('wm_users', 'WHERE id ='. sql_num($$k{user_id}),
						'reply_include' 	=> sql_bool($in->{reply}),
						'true_delete' 		=> sql_bool($in->{delete}),
						'session_length'	=> sql_num($session),
						'fldr_showcount'	=> sql_num($in->{count}),
						'fldr_sortorder'	=> sql_num($in->{order}),
						'fldr_sortfield'	=> sql_str($in->{field})));

	db_commit($$k{dbh});
		
	return($k->_relocate($r, $$k{rootp}));
} # END $k->do_edit

#-------------------------------------------------
# $k->do_main($r)
#-------------------------------------------------
sub do_main {
	my ($k, $r) = @_;

	$$k{page_title} .= 'Preferences';

	return(ht_div({ 'class' => 'box' }),

			ht_table(),
			ht_tr(undef,
				ht_td({ 'class' => 'hdr' }, "Preferences for $$k{user}"),
				ht_td({ 'class' => 'rhdr' },
					'[', ht_a("$$k{rootp}/edit", 'Update'), ']')),
		
			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Reply Included'),
				ht_td(undef, ($$k{p_reply} ? 'Yes' : 'No'))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Delete Action'),
				ht_td(undef, ($$k{p_delete} ? 'Delete' : 'Move to trash'))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Session Length'),
				ht_td(undef, $k->s2hm($$k{p_sess_s}))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Message Count'),
				ht_td(undef, $$k{p_fcount})),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Sort Field'),
				ht_td(undef, ucfirst($$k{p_sfield}))),

			ht_tr(undef,
				ht_td({ 'class' => 'shd' }, 'Folder Sort Order'),
				ht_td(undef, ($$k{p_sorder} ? 'Descending' : 'Ascending'))),

			ht_utable(),
			ht_udiv());
} # END $k->do_main

# EOF
1;

__END__

=head1 NAME 

Alchemy::WebMail::Preferences - Preferences

=head1 SYNOPSIS

  use Alchemy::WebMail::Preferences;

=head1 DESCRIPTION

This module allows users to manipulate their preferences for the
application, these are set to the application wide defaults on the users
first login. 

=head1 APACHE

This is a sample of the location required for this module to run.
Consult Alchemy::WebMail(3) to learn about the configuration options.

  <Location /webmail/mail/prefs >
    SetHandler  perl-script

    PerlHandler Alchemy::WebMail::Preferences
  </Location>

=head1 DATABASE

This is the core table that this module manipulates.

  create table "wm_users" (
    id                  int4 PRIMARY KEY DEFAULT NEXTVAL( 'wm_users_seq' ),
    reply_include       bool,       /* include original in reply */
    session_length      int2,       /* length of users session */
    fldr_showcount      int2,       /* number of emails in folder to show */
    fldr_sortorder      int2,       /* folder sort order */
    fldr_sortfield      varchar,    /* folder sort field */
    username            varchar     /* imap user name */
  );

=head1 SEE ALSO

Alchemy::Webmail(3), Alchemy(3), KrKit(3)

=head1 LIMITATIONS

=head1 AUTHOR

Nicholas Studt <nicholas@photodwarf.org>

=head1 COPYRIGHT

Copyright (c) 2003-2010 Nicholas Studt. All rights reserved.

You may distribute under the terms of either the GNU General Public
License or the Artistic License, as specified in the Perl README file.

=cut
